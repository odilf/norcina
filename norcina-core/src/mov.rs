use std::fmt;

use enum_dispatch::enum_dispatch;

#[enum_dispatch]
pub trait Move {}

#[enum_dispatch]
pub trait InvertibleMove: Move {
    /// The move that undoes `self`.
    fn inverse(&self) -> Self;
}

#[enum_dispatch]
pub trait RandomMove: Move {
    /// Generates a random move.
    fn random(rng: &mut impl rand::Rng) -> Self;
}

#[enum_dispatch]
pub trait DisplayMove: Move + fmt::Display {}
impl<M: Move + fmt::Display> DisplayMove for M {}
