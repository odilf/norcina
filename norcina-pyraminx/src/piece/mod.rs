mod center;
pub mod edge;
mod tip;

use std::fmt;

pub use center::Centers;
pub use edge::Edge;
use norcina_core::types::{Axis, Direction, Orientation3};
pub use tip::Tips;

/// One of the 4 "vertices" of the Pyraminx, which give the names to the moves.
///
/// # Invariants
/// Bitpattern is always either 0, 1, 2 or 3.
#[repr(transparent)]
#[derive(Clone, Copy, PartialEq, Eq)]
pub struct Vertex {
    // Packed field: ------yx
    data: u8,
}

impl Vertex {
    #[inline(always)]
    pub const fn u8(self) -> u8 {
        self.data
    }

    #[inline(always)]
    pub const fn from_u8(data: u8) -> Self {
        assert!(data < 4);
        Self { data }
    }

    #[inline(always)]
    pub const fn from_u8_mod4(data: u8) -> Self {
        Self { data: data & 0b11 }
    }

    #[inline(always)]
    fn from_i8_mod4(n: i8) -> Vertex {
        Self {
            data: n.rem_euclid(4) as u8,
        }
    }

    pub const ALL: [Self; 4] = [Self::U, Self::L, Self::B, Self::R];

    //                                 zy
    pub const U: Self = Self { data: 0b00 };
    pub const L: Self = Self { data: 0b01 };
    pub const B: Self = Self { data: 0b10 };
    pub const R: Self = Self { data: 0b11 };

    #[inline(always)]
    pub const fn x(self) -> Direction {
        Direction::from_u8_any((self.data ^ (self.data >> 1)) & 0b1)
    }

    #[inline(always)]
    pub const fn y(self) -> Direction {
        Direction::from_u8_mod2(self.data)
    }

    #[inline(always)]
    pub const fn z(self) -> Direction {
        Direction::from_u8_any(self.data & 0b10)
    }

    #[inline(always)]
    pub const fn direction_on_axis(self, axis: Axis) -> Direction {
        //        -------zy-------    -----------------x---------------
        let zyx = (self.data << 1) ^ (self.data & 0b1) ^ (self.data >> 1);
        Direction::from_u8_any(zyx & (1 << axis.u8()))
    }

    /// Whether the piece is oriented when the piece is looking towards the move
    /// corresponding with this vertex, or towards the opposite side. In other
    /// words, positive for R and U, negative for L and B.
    ///
    /// See also [`Edge`](Edge#orientation)
    #[inline(always)]
    pub fn orientation_direction(self) -> Direction {
        self.x()
    }

    /// Finds the vertex that is the desired orientation offseted from `start`
    /// around `self`. That is, the vertex that is `orientation` clockwise turns
    /// away from `start` around `self`.
    ///
    /// For example,
    /// - `offset(R, U, 0) == U`
    /// - `offset(R, U, 1) == B`
    /// - `offset(R, U, 2) == L`
    ///
    /// To spell it out, all cycles are:
    ///
    /// ```text
    /// Base |       Cycle
    ///   R  | [L -> U -> B -> L]
    ///   L  | [U -> R -> B -> U]
    ///   B  | [R -> U -> L -> R]
    ///   U  | [R -> L -> B -> R]
    /// ```
    ///
    /// # Panics
    ///
    /// If `self == start`
    #[inline(always)]
    pub fn offset(self, start: Vertex, orientation: Orientation3) -> Self {
        let anchor = self.u8();
        let start = start.u8();
        assert_ne!(anchor, start);

        // If you inspect the table, you see that if the first bit is 0 you need
        // to subtract one to the vertex, otherwise you add one.
        let dont_negate = (anchor as i8 ^ 0b01) & 0b01;
        let sign = dont_negate ^ (dont_negate - 1);

        // From https://graphics.stanford.edu/~seander/bithacks.html#ConditionalNegate
        // Subtract if 0, add if 1.
        let diff = sign * orientation.u8() as i8;

        // In principle, `output = start + diff`. However, we want to skip `anchor` as a value if we were to hit it.
        // We first move to a space where the anchor is always 3. Then we can do
        // mod3 to avoid the anchor. Then we move back to the normal space.

        // m3s = mod 3 space
        let start_m3s = (start + (3 - anchor)) % 4;
        // Diff is relative, rotation of space doesn't affect it.
        let end_m3s = (start_m3s as i8 + diff).rem_euclid(3);
        let end = end_m3s - (3 - anchor as i8);

        Vertex::from_i8_mod4(end)

        // NOTE: An alternative approach, since orientation is either 0b00,
        // 0b01 or 0b10, is to switch the first two bits of the orientation,
        // but doing it conditionally is a bit trickier, probably not even that
        // different.
    }

    /// Offsets a face in the same way as [`Self::offset`].
    ///
    /// Although semantics are kind of differnet, the implementation is
    /// literally just a [`Self::offset`] wrapper.
    pub fn offset_face(self, face: Face, orientation: Orientation3) -> Face {
        Face {
            vertex: self.offset(face.vertex, orientation),
        }
    }
}

impl fmt::Debug for Vertex {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "Vertex {{ {} }}",
            match *self {
                Self::R => 'R',
                Self::L => 'L',
                Self::B => 'B',
                Self::U => 'U',
                _ => unreachable!(),
            }
        )
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Face {
    // The vertex opposite to the given face.
    pub vertex: Vertex,
}

pub type Sticker = Face;

impl Face {
    pub const R: Face = Face { vertex: Vertex::R };
    pub const L: Face = Face { vertex: Vertex::L };
    pub const B: Face = Face { vertex: Vertex::B };
    pub const U: Face = Face { vertex: Vertex::U };

    pub const fn new(opposite_vertex: Vertex) -> Self {
        Self {
            vertex: opposite_vertex,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn fn_vertex_x_logic_is_correct() {
        for (input, output) in [(0b00, 0), (0b01, 1), (0b10, 1), (0b11, 0)] {
            assert_eq!(Vertex { data: input }.x().u8(), output)
        }
    }

    #[test]
    fn fn_offset_has_correct_results_for_every_input() {
        let r = Vertex::R;
        let l = Vertex::L;
        let u = Vertex::U;
        let b = Vertex::B;

        for (anchor, cycle) in [
            (u, [r, l, b]),
            (r, [l, u, b]),
            (l, [u, r, b]),
            (b, [r, u, l]),
        ] {
            for start in 0..3 {
                for i in 0..3 {
                    let orientation = Orientation3::from_u8(i as u8);
                    assert_eq!(
                        anchor.offset(cycle[start], orientation),
                        cycle[(start + i) % 3],
                        "Expected anchor {anchor:?} with face {2:?} offseted by {orientation:?} to result in {1:?} (from cycle {cycle:?}), but it resulted in {0:?}",
                        anchor.offset(cycle[start], orientation),
                        cycle[(start + i) % 3],
                        cycle[start],
                    );
                }
            }
        }
    }

    #[test]
    fn rotating_either_of_a_pair_of_vertices_around_the_other_results_in_the_same_vertex() {
        for v1 in [Vertex::R, Vertex::L, Vertex::U, Vertex::B] {
            for v2 in [Vertex::R, Vertex::L, Vertex::U, Vertex::B] {
                if v1 == v2 {
                    continue;
                }

                for offset in [Orientation3::ONE, Orientation3::TWO] {
                    assert_eq!(v1.offset(v2, offset), v2.offset(v1, -offset))
                }
            }
        }
    }
}
