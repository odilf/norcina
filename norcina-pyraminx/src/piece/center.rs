use super::Vertex;
use crate::mov::CoreMove;
use norcina_core::types::Orientation3;

#[repr(transparent)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Centers {
    // Packed struct: BBLLUURR
    data: u8,
}

impl Centers {
    pub const SOLVED: Self = Self { data: 0 };

    /// The orientation of the tip at the given vertex, relative to the
    /// corresponding center.
    ///
    /// 0 indicates solved, 1 indicates it's one clockwise rotation from
    /// solved, 2 indicates two clockwise rotations (or, equivalently, one
    /// counterclockwise rotation).
    ///
    /// That is, every clockwise move increases orientation by 1.
    #[inline(always)]
    pub const fn orientation_of(self, vertex: Vertex) -> Orientation3 {
        Orientation3::from_u8((self.data >> (vertex.u8() * 2)) & 0b11)
    }

    // TODO: Make const
    pub fn mov(self, mov: CoreMove) -> Self {
        let orientation_delta = mov.amount().u8() + 1;
        let new_orientation = ((self.data >> (mov.vertex().u8() * 2)) + orientation_delta) % 3;

        // TODO: Is there a more consice way to overwrite using bit twiddling?
        Self {
            data: self.data
                // Clear the old bits
                & (!(0b11 << (mov.vertex().u8() * 2)))
                // Fill in with new bits
                | (new_orientation << (mov.vertex().u8() * 2)),
        }
    }
}
