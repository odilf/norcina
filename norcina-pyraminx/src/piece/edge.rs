use norcina_core::types::{Axis, Direction};

use crate::mov::{Amount, CoreMove};

use super::{Face, Vertex};

/// The main pieces of a Pyraminx. There are 6 of them.
///
/// Edge positions correspond to 3x3 centers, in fact. There's six of them in
/// both, and if you inscribe a Pyraminx inside a 3x3 the Pyraminx edges exactly
/// overlap with the 3x3's centers.
///
/// # Orientation
///
/// They also have an orientation. Two of the moves toggle orientation, the
/// other two don't. This decision is arbitrary, so let's arbitrarily choose the
/// moves not on the +x axis, i.e., L and B moves toggle orientation.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Edge {
    // Packed field: ----oaad.
    //
    // d: direction
    // aa: axis
    // o: orientation
    data: u8,
}

impl Edge {
    #[inline(always)]
    pub const fn solved(axis: Axis, direction: Direction) -> Self {
        Self {
            data: direction.u8() + (axis.u8() << 2),
        }
    }

    pub const SOLVED: [Self; 6] = [
        Self::from_index(0),
        Self::from_index(1),
        Self::from_index(2),
        Self::from_index(3),
        Self::from_index(4),
        Self::from_index(5),
    ];

    #[inline(always)]
    pub const fn from_index(index: u8) -> Self {
        Self { data: index }
    }

    #[inline(always)]
    pub const fn u8(self) -> u8 {
        self.data
    }

    #[inline(always)]
    pub const fn axis_direction(self) -> Direction {
        Direction::from_u8_mod2(self.u8())
    }

    #[inline(always)]
    pub const fn axis(self) -> Axis {
        Axis::from_u8((self.u8() >> 1) & 0b11)
    }

    pub const fn is_oriented(self) -> bool {
        self.u8() >> 3 == 0
    }

    pub fn position_from_faces([f1, f2]: [Face; 2]) -> Self {
        //                     Shared
        // xor   00 01 10 11     axis 00 01 10 11
        //
        //  00   00 01 10 11     00   -- 10 01 00
        //  01   -- 00 11 10     01   -- -- 00 01
        //  10   -- -- 00 01     10   -- -- -- 10
        //  11   -- -- -- 00     11   -- -- -- --
        //
        // In summary, you xor and flip. Guess what that is: <->. Makes sense,
        // since we're checking equality. The thing that turns out particularly
        // nicely somewhat coincidentally is that we don't need to transform
        // the bitpattern any further because if the y bit matches, we get 01
        // which is the y-axis index, same for the z bit, 10 and the z-axis and
        // finally the x axis only appears if both are different, for which we
        // get 00 which is, again, the x-axis index!
        let shared_axis = Axis::from_u8_mod3(f1.vertex.u8() ^ f2.vertex.u8() ^ 0b11);

        // We flip the direction because the position of the edge is opposite to
        // the vertices that define their faces.
        let direction = f1.vertex.direction_on_axis(shared_axis).u8() ^ 0b1;
        debug_assert_eq!(
            f1.vertex.direction_on_axis(shared_axis),
            f2.vertex.direction_on_axis(shared_axis),
            "f1={f1:?}, f2={f2:?}, shared_axis={shared_axis:?}"
        );
        Self {
            data: direction + (shared_axis.u8() << 1),
        }
    }

    /// The two vertices this edge "connecting" (even though it doesn't really
    /// touch the vertex itself)
    pub fn vertices(self) -> [Vertex; 2] {
        let mut v1 = 0b00;
        let mut v2 = 0b11;

        let d = self.axis_direction().u8();
        // We xor with the direction, because, for the x-axis, if direction is
        // 0 the coordinates we want are 00 and 11, otherwise we want 01 and
        // 10. For the other axes we overwrite it so it doesn't matter. But this
        // is branchless!
        v1 ^= d;
        v2 ^= d;

        let a = self.axis().u8();

        // If the axis is not x, overwrite the data in that axis by the axis direction.
        //        clear   clear if d is 0.
        v1 = v1 & (!((1 << a) >> 1)) | ((d << a) >> 1);
        v2 = v2 & (!((1 << a) >> 1)) | ((d << a) >> 1);

        [Vertex::from_u8(v1), Vertex::from_u8(v2)]
    }

    /// Faces that make up this edge.
    ///
    /// The opposite of [`Edge::vertices`].
    pub fn faces(self) -> [Face; 2] {
        // Exact same implementation as `Edge::vertices`, except we flip the `d`.
        let mut v1 = 0b00;
        let mut v2 = 0b11;

        let d = self.axis_direction().u8() ^ 0b1;
        v1 ^= d;
        v2 ^= d;

        let a = self.axis().u8();

        v1 = v1 & (!((1 << a) >> 1)) | ((d << a) >> 1);
        v2 = v2 & (!((1 << a) >> 1)) | ((d << a) >> 1);

        [
            Face::new(Vertex::from_u8(v1)),
            Face::new(Vertex::from_u8(v2)),
        ]
    }

    /// Either the U face or the face in the clockwise U direction.
    pub fn orientation_face(self) -> Face {
        // An edge has a U face if the axis direction is negative
        if self.axis_direction() == Direction::Negative {
            Face::U
        } else {
            // Otherwise, depends
            // Axis:      Numerically:
            // X -> B  |  0 -> 2
            // Y -> L  |  1 -> 1
            // Z -> R  |  2 -> 3
            //
            // (4 - axis) % 3 + 1. Long story
            Face {
                vertex: Vertex::from_u8((4 - self.axis().u8()) % 3 + 1),
            }
        }
    }

    pub fn non_orientation_face(self) -> Face {
        // Similar to `Self::orientation_face`, negative => has a U face
        if self.axis_direction() == Direction::Negative {
            // This time, the non-U face depends on the axis too.
            // Namely:    And numerically:
            // X -> R     0 -> 3
            // Y -> B     1 -> 2
            // Z -> L     2 -> 1
            //
            // 3 - axis
            Face {
                vertex: Vertex::from_u8(3 - self.axis().u8()),
            }
        } else {
            // Also depends,
            // axis:       Numerically:
            // X -> L      00 -> 01
            // Y -> R      01 -> 11
            // Z -> B      10 -> 10
            //
            // Swap bits + 1
            let a = self.axis().u8();
            let swapped = ((a >> 1) | (a << 1)) & 0b11;
            Face {
                vertex: Vertex::from_u8(swapped + 1),
            }
        }
    }

    /// Whether the piece is on the U face.
    pub fn is_on_orientation_face(self) -> bool {
        self.axis_direction() == Direction::Negative
    }
}

pub fn move_pieces(mut edges: [Edge; 6], mov: CoreMove) -> [Edge; 6] {
    move_pieces_in_place(&mut edges, mov);
    edges
}

pub fn move_pieces_in_place(edges: &mut [Edge; 6], mov: CoreMove) {
    // TODO: Maybe we could have `mov.vertex_axis()` functions to be more
    // efficient with unpacking and stuff.
    let [a, b, c] = [
        (mov.vertex().x().u8() + (Axis::X.u8() << 1)) as usize,
        (mov.vertex().y().u8() + (Axis::Y.u8() << 1)) as usize,
        (mov.vertex().z().u8() + (Axis::Z.u8() << 1)) as usize,
    ];

    // TODO: Try to do branchless
    if mov.amount() == Amount::Reverse {
        // TODO: Probably can be more efficient.
        edges[a].data ^= (Edge::from_index(a as u8).is_on_orientation_face() as u8) << 3;
        edges[b].data ^= (Edge::from_index(b as u8).is_on_orientation_face() as u8) << 3;
        edges[c].data ^= (Edge::from_index(c as u8).is_on_orientation_face() as u8) << 3;
    }

    // TODO: Would be nice to remove branching.
    if mov.amount() == Amount::Single {
        edges.swap(a, b);
        edges.swap(b, c);
    } else {
        edges.swap(b, c);
        edges.swap(a, b);
    }

    // Toggle orientations if the piece moves into the U face.
    // TODO: Try to do branchless
    if mov.amount() == Amount::Single {
        // TODO: Probably can be more efficient.
        edges[a].data ^= (Edge::from_index(a as u8).is_on_orientation_face() as u8) << 3;
        edges[b].data ^= (Edge::from_index(b as u8).is_on_orientation_face() as u8) << 3;
        edges[c].data ^= (Edge::from_index(c as u8).is_on_orientation_face() as u8) << 3;
    }
}
