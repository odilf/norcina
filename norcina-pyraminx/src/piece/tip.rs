use super::{Centers, Vertex};
use crate::mov::CoreMove;
use norcina_core::types::Orientation3;

/// Tips hold literally the same information as tips.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Tips(Centers);

impl Tips {
    pub const SOLVED: Self = Self(Centers::SOLVED);

    /// See [`Centers::orientation_of`]
    #[inline(always)]
    pub const fn orientation_of(self, vertex: Vertex) -> Orientation3 {
        self.0.orientation_of(vertex)
    }

    /// See [`Centers::mov`]
    #[inline(always)]
    pub fn mov(self, mov: CoreMove) -> Self {
        Self(self.0.mov(mov))
    }

    #[inline(always)]
    pub fn as_centers(self) -> Centers {
        self.0
    }
}
