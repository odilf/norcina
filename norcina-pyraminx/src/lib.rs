use mov::{CoreMove, Move};
use norcina_core::types::Orientation3;
use owo_colors::{OwoColorize as _, Rgb};
use piece::{Centers, Edge, Face, Tips, Vertex, edge};
use std::fmt;

pub mod mov;
pub mod piece;

/// A [Pyraminx](https://www.worldcubeassociation.org/results/rankings/pyram/)
///
/// See also [`CorePyraminx`].
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Pyraminx {
    core: CorePyraminx,
    tips: Tips,
}

impl Pyraminx {
    pub const SOLVED: Self = Self {
        core: CorePyraminx::SOLVED,
        tips: Tips::SOLVED,
    };

    pub fn mov(mut self, moves: impl IntoIterator<Item = Move>) -> Self {
        for mov in moves {
            self = self.mov_single(mov)
        }
        self
    }

    pub fn mov_single(self, mov: Move) -> Self {
        if mov.is_tip_move() {
            Self {
                tips: self.tips.mov(mov.core()),
                ..self
            }
        } else {
            Self {
                core: self.core.mov(mov.core()),
                tips: self.tips.mov(mov.core()),
            }
        }
    }

    pub fn write(
        self,
        f: &mut fmt::Formatter<'_>,
        color_scheme: ColorScheme,
        render_as_triangles: bool,
    ) -> fmt::Result {
        let mut i = 0;
        let mut write = |f: &mut fmt::Formatter<'_>, sticker| {
            if render_as_triangles {
                i += 1;
                if i % 2 != 0 {
                    write!(f, "{}", "▲".color((color_scheme)(sticker)))
                } else {
                    write!(f, "{}", "▼".color((color_scheme)(sticker)))
                }
            } else {
                write!(f, "{}", "██".color((color_scheme)(sticker)))
            }
        };

        let space = |f: &mut fmt::Formatter<'_>| {
            if render_as_triangles {
                write!(f, "{}", " ")
            } else {
                write!(f, "{}", "  ")
            }
        };

        let smallspace = |f: &mut fmt::Formatter<'_>| {
            if render_as_triangles {
                write!(f, "{}", " ")
            } else {
                write!(f, "{}", "  ")
            }
        };

        // RLU is the front face.

        for row in 0..3 {
            for _ in 0..row {
                space(f)?;
            }
            let row_rev = 2 - row;

            for index in 0..=2 * row_rev {
                write(f, self.sticker_at(Face::R, Vertex::L, row_rev, index))?;
            }

            smallspace(f)?;
            for index in (0..=2 * row).rev() {
                write(f, self.sticker_at(Face::B, Vertex::U, row, index))?;
            }

            smallspace(f)?;
            for index in 0..=2 * row_rev {
                write(f, self.sticker_at(Face::L, Vertex::R, row_rev, index))?;
            }

            writeln!(f)?;
        }

        for row in (0..3).rev() {
            for _ in 0..(6 - row) {
                space(f)?;
            }

            for index in 0..=2 * row {
                write(f, self.sticker_at(Face::U, Vertex::B, row, index))?;
            }

            writeln!(f)?;
        }

        Ok(())
    }

    fn sticker_at(&self, query_face: Face, query_base: Vertex, row: u8, index: u8) -> Face {
        debug_assert!(index <= row * 2);

        // Tips at: (0, 0), (2, 0) or (2, 4)
        // Centers at: (1, 1), (2, 1), (2, 3)
        // Edges at: (1, 0), (1, 2), (2, 2)
        //
        // In a square:
        // row/index  0 1 2 3 4
        //     2      T C E C T
        //     1      E C E
        //     0      T
        //
        // Center if index is odd.
        //
        // Without centers:
        // row/index  0 2 4
        //     2      T E T
        //     1      E E
        //     0      T
        //
        // Edge if either if index == 2 or row == 1

        let is_center = index % 2 != 0;
        let is_edge_or_center = index == 2 || row == 1;
        // We start off with edges, because centers and tips are basically identical.
        if is_edge_or_center && !is_center {
            // We have a face
            // B---C
            //  \ /
            //   A
            // We want to get the vertices that touch the edge.

            // We first get the "index"
            //  --.2--    0 -> (1, 0)    mapping function: f(r, i) = r + i/2 - 1
            //  \    /    1 -> (1, 2)
            //   0  1     2 -> (2, 2)
            //    \/
            let i = row + index / 2 - 1;

            // Get the "other" vertex
            let other_vertex = query_face
                .vertex
                .offset(query_base, Orientation3::from_u8_mod3(i + 1));
            let position_faces = [query_face, Face::new(other_vertex)];

            // TODO: Maybe make a proper `EdgePosition` struct...
            let edge_position = Edge::position_from_faces(position_faces);
            let edge = self.core.edges[edge_position.u8() as usize];

            if (query_face == edge_position.orientation_face()) == edge.is_oriented() {
                edge.orientation_face()
            } else {
                edge.non_orientation_face()
            }
        } else {
            let is_center = index % 2 != 0;
            let (i, set) = if is_center {
                (row.max(index) - 1, self.core.centers)
            } else {
                (row / 2 + index / 4, self.tips.as_centers())
            };

            // Find the relevant vertex.
            let vertex = query_face
                .vertex
                .offset(query_base, -Orientation3::from_u8(i));

            // Find the color by "undoing" the given rotation.
            vertex.offset_face(query_face, -set.orientation_of(vertex))
        }
    }
}

/// The parts that are "interesting" of a [`Pyraminx`]. That is, the
/// [`Pyraminx`] without the tips.
///
/// This is because solving (or manipulating in any way) the tips is trivial:
/// Every tip is independent from each other and they can be in one of 3 states
/// without affecting each other. The number of states is not thaaat low (3⁴ =
/// 81) but it's extremely easy to basically map out all states, because the
/// transition between them are, again, very straightforward.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct CorePyraminx {
    centers: Centers,
    edges: [Edge; 6],
}

impl CorePyraminx {
    pub const SOLVED: Self = Self {
        centers: Centers::SOLVED,
        edges: Edge::SOLVED,
    };

    pub fn mov(self, mov: CoreMove) -> Self {
        Self {
            centers: self.centers.mov(mov),
            edges: edge::move_pieces(self.edges, mov),
        }
    }
}

pub type ColorScheme = fn(Face) -> Rgb;

pub const DEFAULT_COLOR_SCHEME: ColorScheme = |sticker| match sticker {
    Face::R => Rgb(214, 59, 39),
    Face::L => Rgb(23, 125, 235),
    Face::B => Rgb(39, 214, 95),
    Face::U => Rgb(235, 231, 23),
    _ => unreachable!(),
};

impl fmt::Display for Pyraminx {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        self.write(f, DEFAULT_COLOR_SCHEME, false)
    }
}

#[cfg(feature = "quickcheck")]
mod quickcheck_impl {
    use quickcheck::{Arbitrary, Gen};
}

#[cfg(test)]
mod tests {
    use mov::Amount;

    use super::*;

    #[test]
    #[cfg_attr(miri, ignore)]
    fn snapshot_solved_pyraminx() {
        insta::assert_snapshot!(Pyraminx::SOLVED)
    }

    #[test]
    #[cfg_attr(miri, ignore)]
    fn snapshot_every_move() {
        for vertex in Vertex::ALL {
            for amount in [Amount::Single, Amount::Reverse] {
                for tip in [false, true] {
                    dbg!(vertex, amount, tip);
                    let mov = Move::new(vertex, amount, tip);
                    insta::assert_snapshot!(
                        format!("Move: {mov} (tip={tip})"),
                        Pyraminx::SOLVED.mov([mov])
                    );
                }
            }
        }
    }

    #[cfg(feature = "quickcheck")]
    quickcheck::quickcheck! {
         // fn move_and_inverse_is_identity(mov: Move, state: Pyraminx) -> bool {
         //     state.mov([mov, mov.inverse()]) == state
         // }

         // fn triple_move_is_identity(mov: Move, state: Pyraminx) -> bool {
         //     state.mov([mov, mov, mov]) == state
         // }

         // fn from_faces_faces_is_identity(f1: Face, f2: Face) -> bool {
         //     let n = Edge::position_from_faces([f1, f2]).faces();
         //     n.contains(&f1) && n.contains(&f2)
         // }
    }
}
