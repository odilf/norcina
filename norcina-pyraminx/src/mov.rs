use std::{fmt, mem};

use crate::piece::Vertex;

// TODO: Maybe make a struct with a u8...
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Amount {
    Single = 0,
    Reverse = 1,
}

impl Amount {
    #[inline(always)]
    pub const fn u8(self) -> u8 {
        self as u8
    }

    /// # Safety
    ///
    /// `data` must be either 0 or 1.
    pub const unsafe fn from_u8_unchecked(data: u8) -> Self {
        debug_assert!(data < 2);
        // SAFETY: Precodnition is that it is either 0 or 1.
        unsafe { mem::transmute(data) }
    }

    pub const fn from_u8(data: u8) -> Self {
        assert!(data < 2);
        // SAFETY: We just checked that `data` is less than 2, so it is either 0 or 1.
        unsafe { mem::transmute(data) }
    }

    pub const fn from_u8_mod2(data: u8) -> Amount {
        // SAFETY: We take modulo 2.
        unsafe { Self::from_u8_unchecked(data & 0b1) }
    }
}

#[repr(transparent)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct CoreMove {
    // Packed field: -----avv
    data: u8,
}

impl CoreMove {
    pub const fn new(vertex: Vertex, amount: Amount) -> Self {
        Self {
            data: vertex.u8() + (amount.u8() << 2),
        }
    }

    pub const ALL: [CoreMove; 8] = [
        CoreMove::new(Vertex::R, Amount::Single),
        CoreMove::new(Vertex::U, Amount::Single),
        CoreMove::new(Vertex::L, Amount::Single),
        CoreMove::new(Vertex::B, Amount::Single),
        CoreMove::new(Vertex::R, Amount::Reverse),
        CoreMove::new(Vertex::U, Amount::Reverse),
        CoreMove::new(Vertex::L, Amount::Reverse),
        CoreMove::new(Vertex::B, Amount::Reverse),
    ];

    #[inline(always)]
    pub const fn vertex(self) -> Vertex {
        Vertex::from_u8_mod4(self.data & 0b011)
    }

    #[inline(always)]
    pub const fn amount(self) -> Amount {
        Amount::from_u8_mod2(self.data >> 2)
    }
}

#[repr(transparent)]
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Move {
    // Packed field: ----tavv
    data: u8,
}

impl Move {
    pub const R: Self = Self::new(Vertex::R, Amount::Single, false);
    pub const RP: Self = Self::new(Vertex::R, Amount::Reverse, false);
    pub const L: Self = Self::new(Vertex::L, Amount::Single, false);
    pub const LP: Self = Self::new(Vertex::L, Amount::Reverse, false);
    pub const B: Self = Self::new(Vertex::B, Amount::Single, false);
    pub const BP: Self = Self::new(Vertex::B, Amount::Reverse, false);
    pub const U: Self = Self::new(Vertex::U, Amount::Single, false);
    pub const UP: Self = Self::new(Vertex::U, Amount::Reverse, false);

    pub const fn new(vertex: Vertex, amount: Amount, tip: bool) -> Self {
        Self {
            data: vertex.u8() + (amount.u8() << 2) + ((tip as u8) << 3),
        }
    }
    pub const fn is_tip_move(self) -> bool {
        (self.data & 0b1000) != 0
    }

    pub const fn core(self) -> CoreMove {
        CoreMove {
            data: self.data & 0b0111,
        }
    }

    pub const fn toggle_tip(self) -> Self {
        Self {
            data: self.data ^ 0b1000,
        }
    }
}

impl fmt::Display for Move {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let addendum = if self.core().amount() == Amount::Single {
            ' '
        } else {
            '\''
        };

        let main = match (self.core().vertex(), self.is_tip_move()) {
            (Vertex::U, false) => "U",
            (Vertex::U, true) => "u",
            (Vertex::L, false) => "L",
            (Vertex::L, true) => "l",
            (Vertex::B, false) => "B",
            (Vertex::B, true) => "b",
            (Vertex::R, false) => "R",
            (Vertex::R, true) => "r",
            (_, _) => unreachable!(),
        };

        write!(f, "{main}{addendum}")
    }
}
