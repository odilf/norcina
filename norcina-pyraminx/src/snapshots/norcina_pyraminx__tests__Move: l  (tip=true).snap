---
source: norcina-pyraminx/src/lib.rs
expression: "Pyraminx::SOLVED.mov([mov])"
---
[38;2;214;59;39mâ–ˆâ–ˆ[39m[38;2;214;59;39mâ–ˆâ–ˆ[39m[38;2;214;59;39mâ–ˆâ–ˆ[39m[38;2;214;59;39mâ–ˆâ–ˆ[39m[38;2;214;59;39mâ–ˆâ–ˆ[39m  [38;2;39;214;95mâ–ˆâ–ˆ[39m  [38;2;23;125;235mâ–ˆâ–ˆ[39m[38;2;23;125;235mâ–ˆâ–ˆ[39m[38;2;23;125;235mâ–ˆâ–ˆ[39m[38;2;23;125;235mâ–ˆâ–ˆ[39m[38;2;23;125;235mâ–ˆâ–ˆ[39m
  [38;2;214;59;39mâ–ˆâ–ˆ[39m[38;2;214;59;39mâ–ˆâ–ˆ[39m[38;2;214;59;39mâ–ˆâ–ˆ[39m  [38;2;39;214;95mâ–ˆâ–ˆ[39m[38;2;39;214;95mâ–ˆâ–ˆ[39m[38;2;39;214;95mâ–ˆâ–ˆ[39m  [38;2;23;125;235mâ–ˆâ–ˆ[39m[38;2;23;125;235mâ–ˆâ–ˆ[39m[38;2;23;125;235mâ–ˆâ–ˆ[39m
    [38;2;235;231;23mâ–ˆâ–ˆ[39m  [38;2;214;59;39mâ–ˆâ–ˆ[39m[38;2;39;214;95mâ–ˆâ–ˆ[39m[38;2;39;214;95mâ–ˆâ–ˆ[39m[38;2;39;214;95mâ–ˆâ–ˆ[39m[38;2;39;214;95mâ–ˆâ–ˆ[39m  [38;2;23;125;235mâ–ˆâ–ˆ[39m
        [38;2;39;214;95mâ–ˆâ–ˆ[39m[38;2;235;231;23mâ–ˆâ–ˆ[39m[38;2;235;231;23mâ–ˆâ–ˆ[39m[38;2;235;231;23mâ–ˆâ–ˆ[39m[38;2;235;231;23mâ–ˆâ–ˆ[39m
          [38;2;235;231;23mâ–ˆâ–ˆ[39m[38;2;235;231;23mâ–ˆâ–ˆ[39m[38;2;235;231;23mâ–ˆâ–ˆ[39m
            [38;2;235;231;23mâ–ˆâ–ˆ[39m
